packages/components/src/checkbox-control/test/index.tsx

```diff
@@ -0,0 +1,105 @@
+/**
+ * External dependencies
+ */
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import BaseCheckboxControl from '..';
+import type { CheckboxControlProps } from '../types';
+
+const getInput = () => screen.getByTestId( 'checkbox' );
+
+const CheckboxControl = ( props: Omit< CheckboxControlProps, 'onChange' > ) => {
+	return (
+		<BaseCheckboxControl
+			onChange={ noop }
+			{ ...props }
+			data-testid="checkbox"
+		/>
+	);
+};
+
+const ControlledCheckboxControl = ( { onChange }: CheckboxControlProps ) => {
+	const [ isChecked, setChecked ] = useState( false );
+	return (
+		<BaseCheckboxControl
+			checked={ isChecked }
+			onChange={ ( value ) => {
+				setChecked( value );
+				onChange( value );
+			} }
+			data-testid="checkbox"
+		/>
+	);
+};
+
+describe( 'CheckboxControl', () => {
+	describe( 'Basic rendering', () => {
+		it( 'should render', () => {
+			render( <CheckboxControl /> );
+			expect( getInput() ).not.toBeNull();
+		} );
+
+		it( 'should render an unchecked `checkbox` by default', () => {
+			render( <CheckboxControl /> );
+			expect( getInput() ).toHaveProperty( 'checked', false );
+		} );
+
+		it( 'should render an checked `checkbox` when `checked={ true }`', () => {
+			render( <CheckboxControl checked /> );
+			expect( getInput() ).toHaveProperty( 'checked', true );
+		} );
+
+		it( 'should render label', () => {
+			render( <CheckboxControl label="Hello" /> );
+
+			const label = screen.getByText( 'Hello' );
+			expect( label ).toBeTruthy();
+		} );
+
+		it( 'should render an indeterminate icon', () => {
+			const { container } = render( <CheckboxControl indeterminate /> );
+
+			const indeterminateIcon = container.getElementsByClassName(
+				'components-checkbox-control__indeterminate'
+			);
```

Comment:
Let's avoid querying the DOM looking for classnames, as we'd be testing implementation details.

What matters to the user is that that the `input` element has an `indeterminate` property with a value of `true`

```tsx
		it( 'should render a checkbox in an indeterminate state', () => {
			render( <CheckboxControl indeterminate /> );
			expect( getInput() ).toHaveProperty( 'indeterminate', true );
		} );
 ```

If we wanted to check also for the icon, we could, for example, use snapshot diffing to target that icon specifically:

```tsx
		it( 'should render the indeterminate icon when in the indeterminate state', () => {
			const { container: containerDefault } = render(
				<CheckboxControl />
			);

			const { container: containerIndeterminate } = render(
				<CheckboxControl indeterminate />
			);

			// Expect the diff snapshot to be mostly about the indeterminate icon
			expect( containerDefault ).toMatchDiffSnapshot(
				containerIndeterminate
			);
		} );
```